File: backend/app/schemas/session.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field
from app.schemas.callback import AgentCurrentState


class TaskConfig(BaseModel):
    """Task configuration."""

    repo_url: str | None = None
    git_branch: str = "main"
    mcp_config: dict = Field(default_factory=dict)
    skill_files: dict = Field(default_factory=dict)


class SessionCreateRequest(BaseModel):
    """Request to create a session."""

    config: TaskConfig | None = None


class SessionUpdateRequest(BaseModel):
    """Request to update a session."""

    status: str | None = None
    sdk_session_id: str | None = None
    workspace_archive_url: str | None = None
    state_patch: dict[str, Any] | None = None
    workspace_files_prefix: str | None = None
    workspace_manifest_key: str | None = None
    workspace_archive_key: str | None = None
    workspace_export_status: str | None = None


class SessionResponse(BaseModel):
    """Session response."""

    session_id: UUID = Field(validation_alias="id")
    user_id: str
    sdk_session_id: str | None
    config_snapshot: dict[str, Any] | None
    workspace_archive_url: str | None
    state_patch: AgentCurrentState | None = None
    workspace_export_status: str | None = None
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class SessionWithTitleResponse(BaseModel):
    """Session response with title (first user prompt).

    @deprecated: Temporary API for frontend development. Will be replaced.
    """

    session_id: UUID = Field(validation_alias="id")
    user_id: str
    sdk_session_id: str | None
    config_snapshot: dict[str, Any] | None
    workspace_archive_url: str | None
    state_patch: AgentCurrentState | None = None
    workspace_export_status: str | None = None
    status: str
    created_at: datetime
    updated_at: datetime
    title: str | None = None

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class SessionStateResponse(BaseModel):
    """Session state response."""

    session_id: UUID = Field(validation_alias="id")
    status: str
    state_patch: AgentCurrentState | None = None
    workspace_export_status: str | None = None
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

File: frontend/features/chat/types/api/session.ts
/**
 * Session-related API types matching backend schemas
 */

import type { ApiStatePatch } from "./callback";

export interface SessionCreateRequest {
  user_id: string;
  config?: TaskConfig | null;
}

export interface SessionUpdateRequest {
  status?: string | null;
  sdk_session_id?: string | null;
  workspace_archive_url?: string | null;
}

export interface SessionResponse {
  session_id: string; // UUID
  user_id: string;
  sdk_session_id: string | null;
  config_snapshot: Record<string, unknown> | null;
  workspace_archive_url: string | null;
  state_patch?: ApiStatePatch | null;
  workspace_export_status?: string | null;
  status: string;
  created_at: string; // ISO datetime
  updated_at: string; // ISO datetime
}

export interface MessageResponse {
  id: number;
  role: string;
  content: Record<string, unknown>;
  text_preview: string | null;
  created_at: string; // ISO datetime
  updated_at: string; // ISO datetime
}

export interface ToolExecutionResponse {
  id: string; // UUID
  message_id: number | null;
  tool_use_id: string | null;
  tool_name: string;
  tool_input: Record<string, unknown> | null;
  tool_output: Record<string, unknown> | null;
  is_error: boolean;
  duration_ms: number | null;
  created_at: string; // ISO datetime
}

export interface UsageResponse {
  total_cost_usd: number | null;
  total_duration_ms: number | null;
  usage_json: Record<string, unknown> | null;
}

export interface TaskConfig {
  repo_url?: string | null;
  git_branch?: string; // defaults to "main"
  mcp_config?: Record<string, unknown>;
  skill_files?: Record<string, unknown>;
}
